packages/router/src/recognize.ts

```diff
@@ -62,51 +63,47 @@ export class Recognizer {
-                // If this implementation changes, update the documentation note in `redirectTo`.
-                this.allowRedirects = false;
-                this.urlTree = e.urlTree;
-                return this.match(e.urlTree);
-              }
-
-              if (e instanceof NoMatch) {
-                throw this.noMatchError(e);
-              }
-
-              throw e;
-            }),
-            map(children => {
-              // Use Object.freeze to prevent readers of the Router state from modifying it outside
-              // of a navigation, resulting in the router being out of sync with the browser.
-              const root = new ActivatedRouteSnapshot(
-                  [], Object.freeze({}), Object.freeze({...this.urlTree.queryParams}),
-                  this.urlTree.fragment, {}, PRIMARY_OUTLET, this.rootComponentType, null, {});
-
-              const rootNode = new TreeNode(root, children);
-              const routeState = new RouterStateSnapshot('', rootNode);
-              const tree = createUrlTreeFromSnapshot(
-                  root, [], this.urlTree.queryParams, this.urlTree.fragment);
-              // https://github.com/angular/angular/issues/47307
-              // Creating the tree stringifies the query params
-              // We don't want to do this here so reassign them to the original.
-              tree.queryParams = this.urlTree.queryParams;
-              routeState.url = this.urlSerializer.serialize(tree);
-              this.inheritParamsAndData(routeState._root);
-              return {state: routeState, tree};
-            }));
+    return this.match(rootSegmentGroup).pipe(map(children => {
+      // Use Object.freeze to prevent readers of the Router state from modifying it outside
+      // of a navigation, resulting in the router being out of sync with the browser.
+      const root = new ActivatedRouteSnapshot(
+          [], Object.freeze({}), Object.freeze({...this.urlTree.queryParams}),
+          this.urlTree.fragment, {}, PRIMARY_OUTLET, this.rootComponentType, null, {});
+
+      const rootNode = new TreeNode(root, children);
+      const routeState = new RouterStateSnapshot('', rootNode);
+      const tree =
+          createUrlTreeFromSnapshot(root, [], this.urlTree.queryParams, this.urlTree.fragment);
+      // https://github.com/angular/angular/issues/47307
+      // Creating the tree stringifies the query params
+      // We don't want to do this here so reassign them to the original.
+      tree.queryParams = this.urlTree.queryParams;
+      routeState.url = this.urlSerializer.serialize(tree);
+      this.inheritParamsAndData(routeState._root);
+      return {state: routeState, tree};
+    }));
   }
 
 
-  private match(tree: UrlTree) {
+  private match(rootSegmentGroup: UrlSegmentGroup): Observable<TreeNode<ActivatedRouteSnapshot>[]> {
+    if (ngDevMode) {
+      currentRedirectCount++;
+      if (currentRedirectCount > 100) {
```

Comment:
nit: should we extract this into a separate const (which would also allow to document this)?
```suggestion
      if (currentRedirectCount > MAX_ALLOWED_REDIRECTS) {
```

Reply:
Yep! Meant to do this but forgot before sending it out for review. Thanks.